src/Model/Query/GeocodeQuery.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+/**
+ * This file is part of the Geocoder package.
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @license    MIT License
+ */
+
+namespace Geocoder\Model\Query;
+
+use Geocoder\Exception\InvalidArgument;
+use Geocoder\Model\Bounds;
+use Geocoder\Model\Coordinates;
+use Geocoder\Provider\Provider;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class GeocodeQuery implements Query
+{
+    /**
+     * @var string
+     */
+    private $text;
+
+    /**
+     * @var Bounds
+     */
+    private $bounds;
+
+    /**
+     * @var string
+     */
+    private $locale;
+
+    /**
+     * @var int
+     */
+    private $limit;
+
+    /**
+     * @var array
+     */
+    private $data;
+
+    /**
+     * @param Coordinates $coordinates
+     */
+    private function __construct($text)
+    {
+        if (empty($text)) {
+            throw new InvalidArgument('Geocode query cannot be empty');
+        }
+
+        $this->text = $text;
+        $this->data = [];
+        $this->limit = Provider::MAX_RESULTS;
+    }
+
+    /**
+     * @param $text
+     *
+     * @return GeocodeQuery
+     */
+    public static function create($text)
+    {
+        return new self($text);
+    }
+
+    /**
+     * @param Bounds $bounds
+     *
+     * @return GeocodeQuery
+     */
+    public function withBounds(Bounds $bounds)
+    {
+        $this->bounds = $bounds;
```

Comment:
> It would not make much sense to have the QueryBuilder immutable, would it?

I'm not that familiar with Doctrine's QB, but as I can see:
* query builder is not perfect match for `GeocoderQuery`, query builder is not a query, it is a service; query, on other hand, is a message, queries should be compared by their values -> identity of query is based on value equality; 
* query builder probably was designed in the mutable way because of performance reasons, because cloning that object is more expensive than small object like `GeocoderQuery`; and number of method calls is much more higher than in your case: for complex query you may trigger `clone` 20-30 times very easy.

However, even for SQL query builder it probably makes sense to keep it immutable too.  Sometimes you need to execute very similar queries, but with little difference (e.g. simple query + the same with `COUNT(*)`). AFAIU, Doctrine suggests to `clone` QB explicitly when you need it, but it looks like a trade-off.

I'd rather compare your query with something like `Request` from PSR-7 if you are in doubt. Request and Query are very similar concepts. Whenever your modify your request/query/command/event, you create different message.